webrisk/database.go

```diff
@@ -0,0 +1,543 @@
+// Copyright 2016 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package webrisk
+
+import (
+	"bytes"
+	"compress/gzip"
+	"context"
+	"encoding/gob"
+	"errors"
+	"log"
+	"math/rand"
+	"os"
+	"sync"
+	"time"
+
+	pb "github.com/GoogleCloudPlatform/golang-samples/webrisk/internal/webrisk_proto"
+	pt "github.com/golang/protobuf/ptypes"
+)
+
+// jitter is the maximum amount of time that we expect an API list update to
+// actually take. We add this time to the update period time to give some
+// leeway before declaring the database as stale.
+const (
+	maxRetryDelay  = 24 * time.Hour
+	baseRetryDelay = 15 * time.Minute
+	jitter         = 30 * time.Second
+)
+
+// database tracks the state of the threat lists published by the Webrisk API.
+// Since the global blacklist is constantly changing, the contents of the
+// database needs to be periodically synced with the Webrisk servers in
+// order to provide protection for the latest threats.
+//
+// The process for updating the database is as follows:
+//	* At startup, if a database file is provided, then load it. If loaded
+//	properly (not corrupted and not stale), then set tfu as the contents.
+//	Otherwise, pull a new threat list from the Web Risk API.
+//	* Periodically, synchronize the database with the Web Risk API.
+//	This uses the Version Token fields to update only parts of the threat list that have
+//	changed since the last sync.
+//	* Anytime tfu is updated, generate a new tfl.
+//
+// The process for querying the database is as follows:
+//	* Check if the requested full hash matches any partial hash in tfl.
+//	If a match is found, return a set of ThreatTypes with a partial match.
+type database struct {
+	config *Config
+
+	// threatsForUpdate maps ThreatTypes to lists of partial hashes.
+	// This data structure is in a format that is easily updated by the API.
+	// It is also the form that is written to disk.
+	tfu threatsForUpdate
+	mu  sync.Mutex // Protects tfu
+
+	// threatsForLookup maps ThreatTypes to sets of partial hashes.
+	// This data structure is in a format that is easily queried.
+	tfl threatsForLookup
+	ml  sync.RWMutex // Protects tfl, err, and last
```

Comment:
Totally minor and feel free to ignore, but it's common to put a group of variables together, with the mutex that protects them at the top.

Reply:
done.